Add support for SOPHGO SoCs and Milk-V boards

xingrz opened the pull request:
This PR ports Zephyr to SOPHGO platforms, including 3 new SoCs and 2 boards: 

| Board | Identifier | SoC |
| ----- | -----------| --- |
| [Milk-V Duo](https://milkv.io/duo) | `milkv_duo/cv1800b/c906_1` | [CV1800B](https://milkv.io/chips/cv1800b) |
| Milk-V Duo (256M variant) | `milkv_duo/sg2002/c906_1` | [SG2002](https://milkv.io/chips/sg2002) |
| [Milk-V Duo S](https://milkv.io/duo-s) | `milkv_duos/sg2002/c906_1` | [SG2000](https://milkv.io/chips/sg2000) |

The CV1800B is a member of the SOPHGO CV180X series. It's a multi-core RV64 SoC, consists with a 1GHz T-Head C906 RV64 big core running Linux, and a 700MHz C906 RV64 little core running RTOS.

The SG2000/SG2002 is a successor of CV1800B, belonging to the CV181x series. It features a 1GHz ARM Cortex-A53 big core, a 1GHz T-Head C906 RV64 big core, and a 700MHz C906 RV64 little core.

This PR uses the `cpuclusters` field introduced by HWMv2 to distinguish between different cores within SOPHGO SoCs. The core names are based on SOPHGO’s official Technical Programming Manual (TPM):

| Identifier | Description |
| -- | -- |
| `a53` | ARM Cortex-A53 big core (available only on SG200x series, selected via BOOT GPIO) |
| `c906_0` | T-Head C906 big core (always available on CV1800B; selected via BOOT GPIO on SG200x) |
| **`c906_1`** | **T-Head C906 little core (the one supported by this PR)** |

This PR brings Zephyr support to the C906 little core (the coprocessor), and lays the groundwork for future support of the A53/C906 big cores.

## TODOs

* [x] HWMv2 ready
* [x] #69650
* [x] #81626
* [x] #81803
* [x] #71773
* [x] #84175
* [x] #89847

xingrz committed:
dts: vendor-prefixes: Add milkv

MilkV is a RISC-V MCU company located in Shenzhen, China.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
dts: vendor-prefixes: Add sophgo

SOPHGO is a company focus on development of TPU and RISC-V processors.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
boards: milkv: Bring up Milk-V Duo

Milk-V Duo is a tiny development board shipped with SOPHGO CV1800B.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
drivers: timer: riscv_machine_timer: Add support for T-Head C906

This commit adapts the existing RISC-V Machine Timer driver to support the
T-Head C906 core, which lacks a standard memory-mapped `MTIME` register.

Changes include:

* Added `mtime-is-32bits` and `mtimecmp-is-32bits` properties to support
  SoCs where `MTIME`/`MTIMECMP` must be accessed as two separate 32-bit
  registers, even on 64-bit platforms.
* Added `mtime-is-csr` property to allow reading the timer value from the
  `time` CSR instead of a memory-mapped `MTIME` register.
* Made the `MTIME` register address optional in DTS; register addresses are
  now resolved by `reg-names` instead of relying on index position.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
drivers: pinctrl: Add pinctrl driver for SOPHGO CVI series

This commit introduces a pin-controller driver for the CVI series SoCs by
SOPHGO.

Predefined pinmux values are also added. Please refer to the official
datasheet for the naming conventions of pins and signals.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
dts: vendor-prefixes: Add thead

T-Head is a semiconductor chip business entity of Alibaba Group.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
dts: bindings: Add T-Head C906 CPU

C906 is a RISC-V CPU designed by T-Head. This commit introduces its DTS
binding for later use.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
drivers: pwm: Add PWM driver for SOPHGO CVI series

This commit adds PWM driver for CVI series SoCs.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
tests: drivers: pwm: Add overlay for Milk-V boards

This commit adds an overlay to the PWM driver tests for Milk-V boards.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

hlleng committed:
drivers: mbox: Add mailbox driver for SOPHGO CVI series

This implements a mailbox driver for CV18xx, a series of multi-core SoCs by
SOPHGO.

This driver enables the little core of CV18xx running RTOS to communicate
with its big core running Linux.

Signed-off-by: honglin leng <a909204013@gmail.com>
Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
dts/soc: sophgo: Add SOPHGO CV1800B

CV180x is a series of RV64 SoCs developed by SOPHGO (formerly CVITEK),
based on T-Head C906 CPU.

Co-authored-by: honglin leng <a909204013@gmail.com>
Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
MAINTAINERS: Add SOPHGO platform and assign myself as maintainer

Also added @lenghonglin as collaborator

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
dts/soc: sophgo: Add SOPHGO SG2000/SG2002

SG2000, also named "CV1813H". SG2002, also named "CV1812CP". Both of them
are multi-core SoCs developed by SOPHGO, belonging to the CV181x series.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz committed:
boards: milkv: Bring up Milk-V Duo S

Duo S is another development board by Milk-V, shipped with SOPHGO SG2000.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

gmarull commented on samples/boards/milkv_duo/mbox/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
```
please, no board specific samples for generic APIs, it makes no sense. Also, sample is broken, please, check how the API changed.

xingrz commented on boards/milkv/duos/milkv_duos.dts:
```diff
@@ -0,0 +1,73 @@
+		};
+	};
+
+	soc {
+		sram: memory@9fe00000 {
+			compatible = "mmio-sram";
+			reg = <0x9fe00000 DT_SIZE_M(2)>;
+		};
+	};
+};
```
In fact, Zephyr is running on the 2nd core of this SoC. The usable RAM region (as well as the start address) of 2nd core is configurable from the 1st core (the main core). This is varant on different particular board of this SoC. The value used here comes from the default configuration of the 1st core Linux for this board: https://github.com/milkv-duo/duo-buildroot-sdk/blob/1d878b3d97517c8fff68ff03194d56b73c21cea5/build/boards/cv181x/cv1813h_milkv_duos_sd/memmap.py#L12-L21

xingrz commented on drivers/gpio/gpio_dw.c:
```diff
@@ -437,10 +437,10 @@ static int gpio_dw_initialize(const struct device *port)
 			    DT_INST_IRQ(n, priority), gpio_dw_isr,				\
 			    DEVICE_DT_INST_GET(n), INST_IRQ_FLAGS(n));				\
-		irq_enable(DT_INST_IRQ_BY_IDX(n, idx, irq));					\
+		irq_enable(DT_INST_IRQN_BY_IDX(n, idx));					\
```
I submit this in a standalone PR: #69650. I'll rebase it as soon as that PR is merged.

fkokosinski commented on boards/milkv/duo256m/doc/index.rst:
```diff
@@ -0,0 +1,81 @@
+.. milkv_duo:
+
+Milk-V Duo 256M
+##########
```
The underline is too short here and in the docs for the other 2 boards, see: https://github.com/zephyrproject-rtos/zephyr/actions/runs/8346902326/job/22845311762?pr=69594

You're also missing an index file, e.g. see this one: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/sifive/index.rst

xingrz commented on dts/riscv/sophgo/cv18xx-c906-common.dtsi:
```diff
@@ -0,0 +1,52 @@
+	};
+
+	soc {
+		compatible = "simple-bus";
```
Added in `dts/riscv/sophgo/cv180x.dtsi` and `dts/riscv/sophgo/cv181x.dtsi`. Thanks for your suggestion.

fkokosinski reviewed (dismissed):
LGTM for the RISC-V parts.

xingrz commented:
Rebased to include the GPIO PR recently merged. I think this PR is ready.

gmarull commented on drivers/pinctrl/pinctrl_sophgo_cvi.c:
```diff
@@ -0,0 +1,37 @@
+
+#define PINCTRL_BASE DT_INST_REG_ADDR(0)
+
+#define PINCTRL_FMUX(base, n) (base + 0x00 + (n) * 4)
```
base is redundant here, driver is a singleton always using instance 0

gmarull commented on drivers/mbox/mbox_sophgo_cvi.c:
```diff
@@ -0,0 +1,139 @@
+static int mbox_cvi_send(const struct device *dev, uint32_t channel, const struct mbox_msg *msg)
+{
+	uint8_t tmp_mbox_info;
+	const struct mbox_cvi_conf *conf = dev->config;
```
suggestion: if not multi-instance, you can also access static config/data directly, as they're not device-dependent.

gmarull commented on dts/riscv/sophgo/cv180x.dtsi:
```diff
@@ -0,0 +1,151 @@
+		mbox: mailbox@1900000 {
+			compatible = "sophgo,cvi-mailbox";
+			reg = <0x1900000 DT_SIZE_K(4)>;
+			interrupt-parent = <&plic0>;
```
why isn't interrupt parent defined ad the soc node? this way you avoid definit it on every child node

gmarull commented on dts/riscv/sophgo/cv180x.dtsi:
```diff
@@ -0,0 +1,151 @@
+
+/ {
+	soc {
+		compatible = "sophgo,cv180x";
```
simple-bus?

xingrz replied:
Do you mean `compatible = "sophgo,cv180x", "simple-bus";`?

gmarull replied:
yes

gmarull commented on dts/riscv/sophgo/cv180x.dtsi:
```diff
@@ -0,0 +1,151 @@
+
+		gpiob: gpio@3021000 {
+			compatible = "snps,designware-gpio";
+			reg = <0x3021000 DT_SIZE_K(4)>;
+			interrupt-parent = <&plic0>;
+			interrupts = <42 1>;
+			gpio-controller;
+			#gpio-cells = <2>;
+			status = "disabled";
+		};
+
+		gpioc: gpio@3022000 {
+			compatible = "snps,designware-gpio";
+			reg = <0x3022000 DT_SIZE_K(4)>;
+			interrupt-parent = <&plic0>;
+			interrupts = <43 1>;
+			gpio-controller;
+			#gpio-cells = <2>;
+			status = "disabled";
+		};
+
+		gpiod: gpio@3023000 {
+			compatible = "snps,designware-gpio";
+			reg = <0x3023000 DT_SIZE_K(4)>;
+			interrupt-parent = <&plic0>;
+			interrupts = <44 1>;
+			gpio-controller;
+			#gpio-cells = <2>;
+			status = "disabled";
+		};
+
+		pwm0: pwm@3060000 {
+			compatible = "sophgo,cvi-pwm";
+			reg = <0x3060000 DT_SIZE_K(4)>;
+			clock-frequency = <DT_FREQ_M(100)>;
+			#pwm-cells = <3>;
+			status = "disabled";
+		};
+
+		pwm1: pwm@3061000 {
```
please, add relevant nodes gradually as you add drivers/support/compatibles. this is throwing the entire DT in one go, which makes it difficult to review.

gmarull commented on soc/sophgo/cv180x/soc.h:
```diff
@@ -0,0 +1,11 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __SOC_H__
+#define __SOC_H__
+
+#include <soc_common.h>
+
+#endif /* __SOC_H__ */
```
why is this needed?

gmarull commented on include/zephyr/dt-bindings/pinctrl/sophgo-cv180x-pinctrl.h:
```diff
@@ -0,0 +1,486 @@
+#define FMUX_SEL_PAD_AUD_AOUTR__XGPIOC_24     3
+#define FMUX_SEL_PAD_AUD_AOUTR__IIS1_DI       4
+#define FMUX_SEL_PAD_AUD_AOUTR__IIS2_DO       5
+#define FMUX_SEL_PAD_AUD_AOUTR__IIS1_DO       6
```
please add all pinctrl related stuff into a single commit

gmarull commented on boards/milkv/duo/doc/index.rst:
```diff
@@ -0,0 +1,82 @@
+After that, you will get a ``fip.bin`` file at the root of the TF card. Use the
+following command to replace the RTOS image with Zephyr:
+
+.. code-block:: sh
+
+   python3 /path/to/duo-buildroot-sdk/fsbl/plat/cv180x/fiptool.py \
+       -v genfip "/path/to/target/fip.bin" \
+       --OLD_FIP="/path/to/original/fip.bin" \
+       --BLCP_2ND="build/zephyr/zephyr.bin"
```
why can't a runner be provided? this looks hacky.

xingrz replied:
Maybe something like `west flash -r cvi --fiptool /path/to/vendor/sdk/fiptool.py --fip-bin /path/to/sdcard/fip.bin --auto-eject`?

gmarull replied:
yes, or rely on `--tool-opt`.

xingrz replied:
Thank you for your suggestion. Update.

gmarull commented on boards/milkv/duo/milkv_duo_defconfig:
```diff
@@ -0,0 +1,10 @@
+
+CONFIG_XIP=n
+
+CONFIG_PINCTRL=y
```
-1, `select PINCTRL` where needed

gmarull commented on boards/milkv/duos/milkv_duos_defconfig:
```diff
@@ -0,0 +1,10 @@
+CONFIG_XIP=n
+
+CONFIG_PINCTRL=y
+CONFIG_SERIAL=y
```
CONFIG_GPIO=y should likely be here as well?

gmarull commented on tests/drivers/pwm/pwm_api/boards/milkv_duo.overlay:
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (c) 2023-2024 Chen Xingyu <hi@xingrz.me>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/ {
+	aliases {
+		pwm-0 = &pwm1;
+	};
+};
+
+&pinctrl {
+	pwm1_default: pwm1_default {
+		group1 {
+			pinmux = <CVI_PINMUX(SD1_D2, PWM_5)>;  /* GP4 */
+		};
+	};
+};
+
+&pwm1 {
+	status = "okay";
+	pinctrl-0 = <&pwm1_default>;
+	pinctrl-names = "default";
+};
```
any reason to not define this at board level?

xingrz replied:
Well, I'm not entirely sure about when it should be defined at the board level versus when it should be an overlay for a test. If I understand correctly, all nodes should be 'disabled' by default and only enabled via overlay when needed, right? So, if all pins are GPIO by default, then PWMs should only be enabled where they are actually used?

gmarull replied:
is this pin meant to always be used for PWM in this board? or is it a random pin chosen for testing only?

xingrz replied:
It's a random pin I choose for testing only.

gmarull reviewed (changes requested):
overall lgtm, but I think needs some better commit organization and a few nitpicks I've seen

xingrz committed:
scripts: runners: Add cvi-fiptool

The SOPHGO CVI series SoCs load the RTOS kernel for the 2nd core from a
`fip.bin` located in the main storage.

This commit introduced a convenient script for updating the `fip.bin` on
the external storage mounted to the host, with the newly built
`zephyr.bin`.

Path to the `fiptool.py` from the official SDK should be provided.

Signed-off-by: Chen Xingyu <hi@xingrz.me>

xingrz commented on drivers/pwm/pwm_sophgo_cvi.c:
```diff
@@ -0,0 +1,143 @@
+};
+
+#define PWM_CVI_INST(n)                                                                            \
+	IF_ENABLED(CONFIG_PINCTRL, (PINCTRL_DT_INST_DEFINE(n);))                                   \
```
This is conditional since there is no need to declare these pinctrl-related structs if pinctrl is not enabled. Of course, this driver works both with and without pinctrl.

gmarull replied:
so where are PWM signals routed when no pinctrl is used?

xingrz replied:
Pinmux can be set from the big core.

xingrz commented:
Hi @gmarull, just wanted to check in on this PR. It's been a couple of weeks since the last update. Is there anything else needed from my side?

xingrz commented:
> @xingrz will you be coming back to this PR, it needs to be rebased

@kartben Rebased. Thanks for reminding.

nordicjm commented on boards/milkv/duo/doc/index.rst:
```diff
@@ -0,0 +1,90 @@
+
+The default configuration can be found in the defconfig file:
+
+        ``boards/milkv/duo/milkv_duo_defconfig``
```
use `:zephyr_file:`

xingrz replied:
Thank you. Fixed.

nordicjm commented on boards/milkv/duo256m/doc/index.rst:
```diff
@@ -0,0 +1,90 @@
+
+The default configuration can be found in the defconfig file:
+
+        ``boards/milkv/duo256m/milkv_duo256m_defconfig``
```
as above

nordicjm commented on boards/milkv/duos/doc/index.rst:
```diff
@@ -0,0 +1,86 @@
+
+The default configuration can be found in the defconfig file:
+
+        ``boards/milkv/duos/milkv_duos_defconfig``
```
as above

nordicjm reviewed (commented):
Looks OK

xingrz commented:
Rebased to resolve merge conflicts.

Is there any chance this could be merged in the next merge window?

nordicjm commented on soc/sophgo/cv180x/Kconfig.defconfig:
```diff
@@ -0,0 +1,25 @@
+	default 25000000
+
+config NUM_IRQS
+	int
```
```suggestion
```

nordicjm replied:
not fixed

nordicjm commented on soc/sophgo/cv180x/Kconfig.defconfig:
```diff
@@ -0,0 +1,25 @@
+
+if SOC_SERIES_CV180X
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 25000000
```
this this needs to come from DTS

nordicjm replied:
@gmarull

xingrz replied:
Thank you for the suggestion. I need some time to investigate the clock tree to understand why the value here (from the official FreeRTOS SDK) is used and why it differs from the CPU frequency stated in the datasheet.

nordicjm replied:
this needs to come from a DTS value

nordicjm replied:
```
clock0: clock@addr {
        compatible = "vendor,device";
        reg = <0xaddr 0xsize>;
        clock-frequency = <25000000>;
};
```
quite literally that

VynDragon replied:
so https://github.com/zephyrproject-rtos/zephyr/issues/37685 ?

nordicjm replied:
dts describes hardware, there is a clock (or multiple) in the device, if you're not describing it in dts then you're omitting required hardware which is needed for the chip to function. You can also reference linux for this e.g. https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/clock/fixed-clock.yaml

VynDragon replied:
That is not helpful, there is no clock driver for this soc, the timebase is not a separate clock, and there is no information available from which to infer the timebase-frequency (and that last bit doesnt apply to bflb because we know where the clock is from and can configure it, but the other issues still apply).

Secondly, I have already referenced linux, where they have a dedicated property for this which the issue i linked references: a timebase-frequency property in the cpu properties: https://github.com/torvalds/linux/blob/master/arch/riscv/boot/dts/sophgo/cv180x-cpus.dtsi, like nandojve proposed years ago, likely for the exact same kind of issues: https://github.com/zephyrproject-rtos/zephyr/issues/37685

nashif replied:
@nordicjm 
SYS_CLOCK_HW_CYCLES_PER_SEC was ALWAYS a kconfig thing

```
git grep CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC | wc -l
285
```

I am not following what are you asking here and what your expections are.

nordicjm replied:
The default value should come from dts, this was raised by Gerard when he was doing dts things and doesn't make sense to have a disjoint value in both, not that someone can't change the Kconfig but the **default** should refer to dts, just as linux does it, just as many SoCs do it

kartben replied:
@nordicjm please revisit in the light of the latest comment

nordicjm replied:
How can it not be stored in dts? That comment makes no sense, if you can store it in Kconfig, you can store it in dts. It doesn't need a driver to have a clock frequency field

nordicjm commented on soc/sophgo/cv181x/Kconfig.defconfig:
```diff
@@ -0,0 +1,25 @@
+
+if SOC_SERIES_CV181X
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 25000000
+
+config NUM_IRQS
+	int
+	default 128
```
as above

nordicjm replied:
not fixed

nordicjm replied:
the comment here refers to a point above about removing `int` not that this value come from dts

nordicjm replied:
it's above that, and shows the same thing as this https://github.com/zephyrproject-rtos/zephyr/pull/69594#discussion_r1845984587

fkokosinski commented:
> Rebased to resolve merge conflicts.
> 
> Is there any chance this could be merged in the next merge window?

In general the changes LGTM (sans things reported by Jamie), but you'd have to fix failing tests first before we can approve the PR.

xingrz commented:
> > Rebased to resolve merge conflicts.
> > Is there any chance this could be merged in the next merge window?
> 
> In general the changes LGTM (sans things reported by Jamie), but you'd have to fix failing tests first before we can approve the PR.

The failing tests seem to be caused by warnings when compiling certain unrelated tests under `CONFIG_64BIT=y` (e.g., passing a `size_t` to `%.*s`). I was able to reproduce the same issue locally with another 64-bit board (`rpi_4b`). Should I open a separate PR to address these warnings?

henrikbrixandersen commented:
> Should I open a separate PR to address these warnings?

Yes, please do.

oresk commented:
I this going to happen? Is there anything I can do to make it happen?

petejohanson commented:
I'm also interested in reviving/helping with this. I've got a Duo S on order and am happy to help support/test this.

0xStuart commented:
Can we get this opened again?  There is a lot of interest in this board.  I also found another attempt to port it.
https://github.com/kinsamanka/milkv-zephyros

petejohanson commented:
I did start the process of rebasing last night. Fixing up the machine timer conflicts, and start tweaking the SoC bits to prep for targeting the different available cores. Targeting the main app core will take a bit of integration for the fsbl they use, if we want to support that.

leiche-cid commented:
Hi, I’m really interested in this PR, especially as it relates to the Milk-V platform, which I find very exciting. Many thanks to @xingrz  for the excellent work!

I was wondering if there’s any update on this PR or if there’s anything I can do to help move it forward?

petejohanson commented:
> Hi, I’m really interested in this PR, especially as it relates to the Milk-V platform, which I find very exciting. Many thanks to @xingrz for the excellent work!
> 
> I was wondering if there’s any update on this PR or if there’s anything I can do to help move it forward?

I'm not sure where @xingrz is in terms of focus on this. I have taken the latest work from them and rebased on current Zephyr `main` and been able to run the co-processor build on my Duo S (with various minor tweaks for OpenAMP resource table support) using the Linux remoteproc setup. I've got some initial legwork done for running Zephyr on the *main* processor (via u-boot using https://lupyuen.github.io/articles/sg2000 as a reference) that I hope to test "soon".

I've manually run a few of the kernel tests on the co-processor and things look good, but I don't have an easy way to automate that to really test it without some custom runner shenanigans.

If @xingrz is interested in continuing their work, I'm happy to offer my rebased/change version for them to integrate here, or I can open a new PR once ready (keep attribution, of course!) to help move this forward.

xingrz commented:
@leiche-cid @petejohanson I’m back now and will catch up on this PR soon 😉

petejohanson commented:
If you're at all interested on the OpenAMP changes, etc. I've been pushing to https://github.com/petejohanson/zephyr/tree/milkv-duo/hwmv2/dev-dev

xingrz commented:
> If you're at all interested in the OpenAMP changes, etc. I've been pushing to https://github.com/petejohanson/zephyr/tree/milkv-duo/hwmv2/dev-dev

@petejohanson Thanks! Do you prefer to open a separate PR for your OpenAMP changes, or should I cherry-pick them into this PR?

petejohanson commented:
> > If you're at all interested in the OpenAMP changes, etc. I've been pushing to https://github.com/petejohanson/zephyr/tree/milkv-duo/hwmv2/dev-dev
> 
> @petejohanson Thanks! Do you prefer to open a separate PR for your OpenAMP changes, or should I cherry-pick them into this PR?

I had rebased them an integrated them into the existing commits of yours. I'm fine doing that as a follow up PR, if we want, once this is in.

petejohanson commented on boards/milkv/duo/milkv_duo.yaml:
```diff
@@ -0,0 +1,14 @@
+identifier: milkv_duo
```
IMHO, all of these boards should drop the `milkv_` prefix for the board IDs, filenames, etc. to match the current convention for these.

Related: Since I'd like to support the main core eventually on these too, I think the initial version should name things accounting for that, e.g. `duos_sg2000_coproc` with an identifier of `duos/sg2000/coproc` to start, so we can target the main core(s) later without having to refactor the existing names.

xingrz replied:
Good idea, just a concern — would names like `duo` or `duos` be a bit too common?

petejohanson replied:
That's a valid question, and one I don't think I've seen address explicitly in the docs. AFAIK, there's no way to specify a vendor/subdirectory when specifying a board ID, so in theory conflicts like that would require putting something like the vendor ID still in the board definition. Maybe others can weigh in.

xingrz replied:
I'd prefer something like `milkv_duo/cv1800b/c906_1` and `milkv_duos/sg2000/a53`.
This matches how SOPHGO directly refers to their cores in their technical manuals.

xingrz replied:
Furthermore, I think we could also merge `milkv_duo` and `milkv_duo256m` into a single board definition, like `milkv_duo/sg2002/c906_1`, since they share the same PCB design, and SOPHGO actually markets the "duo256m" (as named in their SDK) as simply the 256MB variant of DUO.

When this PR was first written, this wasn’t feasible — but now that HWMv2 has landed in mainline, it should be straightforward to support.

petejohanson replied:
Seems very reasonable to me. Pinout looks very similar, a few exceptions, e.g. the LED pin, but given the form factor matches, I think that's a good call.

petejohanson commented on boards/milkv/duo256m/doc/index.rst:
```diff
@@ -0,0 +1,89 @@
+==================
+The Milk-V Duo 256M board configuration supports the following hardware features:
+
+.. list-table::
+   :header-rows: 1
+
+   * - Peripheral
+     - Kconfig option
+     - Devicetree compatible
+   * - Mailbox
+     - :kconfig:option:`CONFIG_MBOX`
+     - :dtcompatible:`sophgo,cvi-mailbox`
+   * - Pin controller
+     - :kconfig:option:`CONFIG_PINCTRL`
+     - :dtcompatible:`sophgo,cvi-pinctrl`
+   * - GPIO
+     - :kconfig:option:`CONFIG_GPIO`
+     - :dtcompatible:`snps,designware-gpio`
+   * - PWM
+     - :kconfig:option:`CONFIG_PWM`
+     - :dtcompatible:`sophgo,cvi-pwm`
+   * - UART
+     - :kconfig:option:`CONFIG_SERIAL`
+     - :dtcompatible:`ns16550`
+   * - PLIC
+     - N/A
+     - :dtcompatible:`sifive,plic-1.0.0`
+   * - SysTick
+     - N/A
+     - :dtcompatible:`thead,machine-timer`
```
This is now auto-generated:

```suggestion
.. zephyr:board-supported-hw::
```

petejohanson reviewed (commented):
A few high level comments.

xingrz commented on drivers/timer/riscv_machine_timer.c:
```diff
@@ -14,14 +14,26 @@
 #include <zephyr/spinlock.h>
 #include <zephyr/irq.h>
 
+#if DT_HAS_COMPAT_STATUS_OKAY(thead_machine_timer)
+#define DT_DRV_COMPAT thead_machine_timer
+
+#define MTIMER_REQUIRES_32BIT_ACCESS
+#define MTIMER_READ_TIME_CSR
+#define MTIMECMP_REG (DT_INST_REG_ADDR(0) + 0x4000U)
+#else
 #define DT_DRV_COMPAT riscv_machine_timer
 
 #define MTIME_REG    DT_INST_REG_ADDR_BY_IDX(0, 0)
 #define MTIMECMP_REG DT_INST_REG_ADDR_BY_IDX(0, 1)
+#endif
```
@VynDragon I'm not sure if I'm doing this right. The issue is that I need a special way to access both the `MTIME` and `MTIMECMP` registers, while the rest of the driver is identical. Could you give me some advice?

VynDragon replied:
Hello, why cant it use the CSRs? I havent ran into that issue for T-Head E907, which works fine with this driver.

VynDragon replied:
Ah no i misunderstand, you only need to set the regs in DTS, MTIME reg and MTIMECMP reg. For example on E907:
```
		mtimer: timer@e000bff8 {
			compatible = "riscv,machine-timer";
			interrupts-extended = <&ictrl 7>;
			reg = <0xE000BFF8 0x8 0xE0004000 0x8>;
			reg-names = "mtime", "mtimecmp";
		};
		```

VynDragon replied:
I looked into the C906 manual and it seems it has no mtime register, only CSR...

VynDragon replied:
I dont see any issue with the current approach, but maintainer opinion is probably a good thing to get:
@fkokosinski @kgugala @tgorochowik

VynDragon replied:
I would only advise signifying from binding name that it is more c906/910 specific than thead specific (e906/7 is fine c920 is fine, etc), seems the read as 32-bit is unavoidable due to endianness and still applies to later cpus though...

VynDragon replied:
I think I have had a better idea, maybe there could be, in the riscv timer binding, 4 optional booleans:
- mtime-is-csr
- mtimecmp-is-csr
- mtime-is-32bits
- mtimecmp-is-32bits

Doing the same job but without a new binding: the -csr ones switch each register to use the register address as a csr read/write instead, and is-32bits, applies to 64-bits archs where mtime or mtimecp are 2 32-bits registers instead of a single 64-bits registers. This would likely be preferred since it takes care of cases in the future and moves the configuration to DTS.

xingrz replied:
I did [a bit of digging](https://github.com/riscv/riscv-isa-manual/blob/main/src/priv-csrs.adoc#csr-listing), and it seems that `time` is indeed a valid CSR — it's defined in the privileged spec as a read-only alias for the `MTIME` register on standard RISC-V CPUs.

However, I don't think we should assume a `timecmp` CSR exists unless we’ve actually seen it implemented in real hardware. So if we go with this approach, I'd suggest being cautious with `mtimecmp-is-csr` unless we have a confirmed case.

xingrz replied:
Updated.

fkokosinski replied:
Thanks!

VynDragon replied:
On the principle this looks good, Thanks!

But what is csr_read(time) meant to mean? I would've expecte csr_read(MTIME_REG).

Also if switching to ADDR_BY_NAME instead of just order, it might needed to updates some dts.

VynDragon replied:
what i mean is the address in MTIME_REG, when using `mtime-is-csr`, would be the CSR's address

xingrz replied:
Ah, I see what you mean now — you're suggesting we directly use the CSR address (e.g. `0xC01`) instead of the named macro like `time`.

I think putting a CSR address like `0xC01` in the DTS `reg` field might be misleading, since it's not a memory address and can't be accessed via load/store. The `reg` field is generally expected to hold memory-mapped addresses, so I'd prefer to leave it out entirely when `mtime-is-csr` is set.

henrikbrixandersen replied:
> Regarding `csr_read(time)`: this expands into the `csrr time` instruction, which is a standard RISC-V assembly instruction (not a memory load). In contrast, `MTIME_REG` refers to a memory-mapped register, accessed via a pointer dereference. These are fundamentally different access mechanisms.
> 
> As far as I know, most standard RISC-V implementations support both — `csrr time` is essentially a shortcut for reading the value pointed to by `MTIME_REG`. But on the C906, `csrr time` is the only way available, since there’s no memory-mapped `MTIME` register at all.
> 
> And yes, I just noticed it did break some CI tests. I think I should clean that up separately in another PR to keep this one focused.

Beware that not all RISC-V implementations support the time[h] CSRs (see e.g. https://stnolting.github.io/neorv32/#_machine_counter_and_timer_csrs).

petejohanson commented on boards/milkv/duo/milkv_duo_cv1800b_c906_1.yaml:
```diff
@@ -0,0 +1,14 @@
+identifier: milkv_duo/cv1800b/c906_1
```
This should probably have the `ram` specified here. Used for testcase filtering.

petejohanson commented on boards/milkv/duo/milkv_duo_sg2002_c906_1.yaml:
```diff
@@ -0,0 +1,14 @@
+identifier: milkv_duo/sg2002/c906_1
```
Ditto

petejohanson reviewed (commented):
A few other thoughts from scanning this again.

xingrz commented on scripts/west_commands/runners/cvi-fiptool.py:
```diff
@@ -0,0 +1,60 @@
+
+    @classmethod
+    def capabilities(cls):
+        return RunnerCaps(commands={'flash'})
```
Yeah, agreed it’s a bit unusual as a runner.

Unlike UF2, which is a self-contained image, FIP embeds Zephyr into a larger binary that also includes things like U-Boot — which developers usually build themselves. So Zephyr can’t fully generate it on its own.

The runner was just a practical way to hook into that flow, but I’m open to better ideas.

petejohanson replied:
Yeah, that's fair. It's a bit of a blurred line with the other deps it's bundled with.

leiche-cid commented:
@VynDragon @fkokosinski @henrikbrixandersen Hi, could you please go back again take a look since all CI checks have passed and all dependencies has merged. I think it's time to move forward.

nordicjm commented on soc/sophgo/cv181x/Kconfig.defconfig:
```diff
@@ -0,0 +1,22 @@
+
+if SOC_SERIES_CV181X
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 25000000
+
```
as above

VynDragon commented:
Of course, I meant past this PR, eg do you intend to continue development for this SoC, or is it only for the sophgo riscv hackaton thing?

VynDragon commented:
@xingrz Do you need any help finishing this?

alexpaschoaletto commented:
Just stumbled upon this topic - I'd just like to say it's awesome to see the milkv-duo receiving attention from the Zephyr community.  Sadly I don't posess the low-level technical knowledge to help with the implementation, but I'll be watching this with great interest. I have been looking for an excuse to buy the Duo for a long time now, and this will certainly be it 😄
